### PR TITLE
Actions: identity

### PR DESCRIPTION
**Pull request by raymondng**
Added support for `SecureSocial`'s `Identity` (for backward compatibility to aid migration to new framework)
`UserRequest`, etc. are now modeled as traits
`buildNonUserRequest` factored out to `UserActionsHelper` to aid injection of different implementations
Added HTML (i.e. `Page`) support
Wired up `OAuth2Controller` (abook import); validated locally

@andrewconner

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 1]
```diff
@@ -5,28 +5,48 @@ import com.keepit.common.db.{ ExternalId, Id }
...
+trait UserRequest[T] extends MaybeUserRequest[T] {
+  def request: Request[T]
+  def userId: Id[User]
+  def adminUserId: Option[Id[User]]
+  def user: User
+  def experiments: Set[ExperimentType]
+}
+
+trait SecureSocialIdentityAccess[T] { self: MaybeUserRequest[T] =>
+  def identityOptF: () => Future[Option[Identity]] = () => Future.successful(None)
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: I'm not sure this is a good pattern. Apologies for not catching it earlier.

``` scala
scala> def t: () => String = () => {println("called"); "done"}
t: () => String

scala> t()
called
res2: String = done

scala> t()
called
res3: String = done

scala> t()
called
res4: String = done
```

Seems like the Future will accidentally get re-generated quite a bit.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 2]
```diff
@@ -5,28 +5,48 @@ import com.keepit.common.db.{ ExternalId, Id }
...
+  lazy val identityOpt = Await.result(identityOptF.apply, 5 seconds)
+}
+
+case class SimpleNonUserRequest[T](request: Request[T]) extends WrappedRequest(request) with NonUserRequest[T] with SecureSocialIdentityAccess[T]
+
+case class SimpleUserRequest[T](
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:36]: Why `Simple` again?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Ah good point. In general the lazy val takes care of caching -- but in this case I didn't separate it out. Will take care of it.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:36]: Yes I reverted back to modeling using traits (extending `Request`) -- so this is now an implementation detail that most clients shouldn't be concerned with (and in fact I might just hide this from them).

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 3]
```diff
@@ -116,6 +137,12 @@ trait UserActions extends Logging { self: Controller =>
     resF.map(maybeAugmentCORS(_))
   }
 
+  private def pageAction[P[_]] = new ActionFunction[P, P] {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Lowercase `pageAction` (others are uppercase)?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 4]
```diff
@@ -116,6 +137,12 @@ trait UserActions extends Logging { self: Controller =>
...
+    def invokeBlock[A](request: P[A], block: (P[A]) => Future[Result]): Future[Result] = {
+      block(request).map(_.withHeaders(CONTENT_TYPE -> HTML))
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: This should be fairly automatic based on what's passed to `Ok` (and etc). Is this a stub for more logic later?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:36]: Will we have non-simple UserRequests?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Nothing other than the fact that it's a `def` not an `object` -- do we name functions with upper-case?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 4 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Actually I find that (e.g. in `TypeaheadAdminController`) it's not 100% automatic so I added explicit support.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 3 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Not usually, but it's used like an object. In the Play doc examples, they use uppercase, so the implementation doesn't matter. May be confusing over time to mix it up.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 4 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: I only see one HTML action there, `index`. Am I missing them?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala [hunk 5]
```diff
@@ -5,28 +5,48 @@ import com.keepit.common.db.{ ExternalId, Id }
...
+  lazy val identityOpt = Await.result(identityOptF.apply, 5 seconds)
+}
+
+case class SimpleNonUserRequest[T](request: Request[T]) extends WrappedRequest(request) with NonUserRequest[T] with SecureSocialIdentityAccess[T]
+
+case class SimpleUserRequest[T](
     request: Request[T],
     userId: Id[User],
     adminUserId: Option[Id[User]],
     userF: () => Future[User],
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:40]: Will we expose the `Future[User]`, or encourage people to just call `request.user`?

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: Sure I was a bit hesitant about this one as well. Will take care of it.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 4 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala]: See the last method (`search`) -- this method is hacky though so I don't blame Play ...

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 5 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:40]: We will, but I also want to see if we can discourage even that.

File: kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/controller/UserActions.scala:36]: Right now we have non-simple `NonUserRequest` and I like the flexibility afforded by the trait-based approach.

**Pull request merged**